### PR TITLE
 feat: Implement next_event and prev_event for SnubaEvent 

### PR DESCRIPTION
**Pull request by lynnagara**
The Snuba implementation also allows searching for a next and previous
event with an environment filter. This will enable navigating through events
in a subset of environments on the issue details page in future.

Required for SEN-238

Depends on #11071

### CODE REVIEW

File: src/sentry/models/event.py [hunk 1]
```diff
@@ -320,75 +310,310 @@ def site(self):
...
+            },
+            orderby=['timestamp', 'event_id'],
+            limit=5
+        )
+
+        if 'error' in result or len(result['data']) == 0:
+            return None
+
+        events = [e for e in result['data'] if
+                  e['timestamp'] == self.timestamp and
```
Comment (alex-hofsteede) [inline src/sentry/models/event.py]: We can just pass these conditions to Snuba and get 1 event. The only reason that we were selecting 5 events and post-filtering them in python before was that Postgres couldn't handle this condition well (too slow) but in snuba we should be able to just specify the condition in the query and get the 1 matching event.

File: src/sentry/models/event.py [hunk 2]
```diff
@@ -320,75 +310,310 @@ def site(self):
...
+        result = snuba.raw_query(
+            start=datetime.utcfromtimestamp(0),  # will be clamped to project retention
+            end=self.datetime,  # lte current event
+            selected_columns=self.selected_columns,
+            conditions=conditions,
+            filter_keys={
+                'project_id': [self.project_id],
+                'issue': [self.group_id],
+            },
+            orderby=['-timestamp', '-event_id'],
```
Comment (alex-hofsteede) [inline src/sentry/models/event.py]: This in-memory ordering could be really expensive still, especially if we are selecting all of the `selected_columns`, basically we have to sort every event from now until the end of time in memory and find the earliest one.

I think I may try and extend the logic in https://github.com/getsentry/snuba/blob/master/snuba/split.py to work with these types of queries. TL;DR it just searches a small time range first, so that if we find something in that range we can avoid searching and sorting every other event.

File: src/sentry/models/event.py [hunk 3]
```diff
@@ -320,75 +310,310 @@ def site(self):
...
+
+        conditions = [['event_id', '!=', self.id]]
+
+        if len(environments) > 0:
+            conditions.append(['environment', 'IN', environments])
+
+        result = snuba.raw_query(
+            start=datetime.utcfromtimestamp(0),  # will be clamped to project retention
+            end=self.datetime,  # lte current event
+            selected_columns=self.selected_columns,
```
Comment (alex-hofsteede) [inline src/sentry/models/event.py]: If I understand correctly, we only need the event_id here to put in the 'next' field. We can return the full set of columns later when the user actually clicks on the next link.  Adding all the `selected columns` here is going to make clickhouse look at a lot more columns than it needs to, and will make this query a lot slower.

**Review by alex-hofsteede — CHANGES_REQUESTED** [review]
Needs some changes to the snuba queries.

Also feel free to rebase on master now as my changes are merged.

File: src/sentry/models/event.py (hunk 1 above)
Comment (lynnagara) [reply src/sentry/models/event.py]: Ah right, updated to drop this and select 1

File: src/sentry/models/event.py (hunk 3 above)
Comment (lynnagara) [reply src/sentry/models/event.py]: Good point, updated these functions to be `next_event_id` and `prev_event_id`

Comment (lynnagara) [reply src/sentry/models/event.py]: It seems the event details endpoint is currently returning the database ids instead of `event_ids` for next/prev id values.
https://github.com/getsentry/sentry/blob/master/src/sentry/api/endpoints/event_details.py#L38-L39

As a result of this we'll also be swapping these out for event_ids there then

Comment (alex-hofsteede) [reply src/sentry/models/event.py]: yeah I think we should move to have the event details endpoint just return `event_id`s for next and prev events, since the details endpoint supports looking up events both ways, it should just work 🤞

File: src/sentry/models/event.py [hunk 4]
```diff
@@ -476,13 +476,57 @@ def id(self):
...
-    @property
-    def next_event(self):
-        return None
+    def next_event_id(self, environments=[]):
+        from sentry.utils import snuba
 
-    @property
-    def prev_event(self):
-        return None
+        conditions = [['event_id', '!=', self.id]]
```
Comment (alex-hofsteede) [inline src/sentry/models/event.py]: I think you still need the `timestamp > self.timestamp OR (timestamp == self.timestamp AND event_id > self.id)` type condition in here.

Comment (lynnagara) [reply src/sentry/models/event.py]: How would you do a string comparison function for event_id like that in a Snuba condition?

Comment (alex-hofsteede) [reply src/sentry/models/event.py]: you can use `>`, `>=` etc on strings, 

The tough part I think is that snuba nested conditions are strung together as an `AND` of `OR`s, and you might need them nested the other way, ie https://github.com/getsentry/snuba#conditions

Comment (alex-hofsteede) [reply src/sentry/models/event.py]: `conditions = [["timestamp", ">=", self.timestamp], [["timestamp", ">", self.timestamp], ["id", ">", self.id]]]` is the above query rearranged to be an AND of ORs. 

and you can definitely do inequality operators on Strings and FixedStrings in clickhouse, so I don't think the comparison should be a problem

Comment (lynnagara) [reply src/sentry/models/event.py]: Thanks, updated 17ed3e1

**Comment by alex-hofsteede** [comment]
OK, so you can now actually rebase on master to get all my SnubaEvent changes

File: src/sentry/api/endpoints/project_event_details.py [hunk 5]
```diff
@@ -41,19 +42,52 @@ def get(self, request, project, event_id):
...
+            return self.get_legacy(request, project, event_id)
+
+        # We might get a database id instead of an event id
+        if not is_event_id(event_id):
+            event = Event.objects.from_event_id(event_id, project.id)
+
+            if event is None:
+                return Response({'detail': 'Event not found'}, status=404)
+
+            event_id = event.event_id
```
Comment (alex-hofsteede) [inline src/sentry/api/endpoints/project_event_details.py]: Is it possible to put this block in some kind of utility function in EventManager? 

Basically something in line with this comment.
https://github.com/getsentry/sentry/blob/master/src/sentry/db/models/manager.py#L355-L356

Basically we should try and make sure everywhere in the code is using `event_id` and if it isn't we can resolve `id` to `event_id` for now (and maybe log the times that this happens so we can fix them?)

File: src/sentry/api/endpoints/project_event_details.py [hunk 6]
```diff
@@ -41,19 +42,52 @@ def get(self, request, project, event_id):
...
+
+        snuba_event = SnubaEvent.get_event(project.id, event_id)
```
Comment (alex-hofsteede) [inline src/sentry/api/endpoints/project_event_details.py]: Maybe this `get_event` should actually be part of the Manager too, like a version of `Event.objects.from_event_id` that returns SnubaEvents instead of old Events?

File: src/sentry/models/event.py [hunk 7]
```diff
@@ -476,13 +476,60 @@ def id(self):
...
+
+        if 'error' in result or len(result['data']) == 0:
+            return None
+
+        return six.text_type(result['data'][0]['event_id'])
+
+    def prev_event_id(self, environments=None):
+        from sentry.utils import snuba
+
+        conditions = [['event_id', '!=', self.id]]
```
Comment (alex-hofsteede) [inline src/sentry/models/event.py]: these should be the same 3 conditions as `next_event_id` but with `<`s instead of `>`s

File: src/sentry/api/endpoints/project_event_details.py (hunk 6 above)
Comment (lynnagara) [reply src/sentry/api/endpoints/project_event_details.py]: Do you mean something like this? 70b3917

**Comment by lynnagara** [comment]
@alex-hofsteede Made those changes, mind taking another look?

File: src/sentry/db/models/manager.py [hunk 8]
```diff
@@ -342,6 +342,30 @@ def bind_nodes(self, object_list, *node_names):
...
+            })
+
+            event = self.from_event_id(id_or_event_id, project_id)
+
+            if not event:
+                return None
+
+            id_or_event_id = event.event_id
+
+        return SnubaEvent.get_event(project_id, id_or_event_id)
```
Comment (alex-hofsteede) [inline src/sentry/db/models/manager.py]: Oops, this is my bad, but its slightly annoying that these params are the opposite way around here. I guess not a big deal.

File: tests/snuba/api/endpoints/test_project_event_details.py [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+        response = self.client.get(url, format='json', data={
+            'enable_snuba': '1',
+            'environment': ['production', 'staging']
+        })
+
+        assert response.status_code == 200, response.content
+        assert response.data['id'] == six.text_type(self.cur_event.event_id)
+        assert response.data['previousEventID'] is None
+        assert response.data['nextEventID'] == self.next_event.event_id
+        assert response.data['groupID'] == six.text_type(self.prev_event.group.id)
```
Comment (alex-hofsteede) [inline tests/snuba/api/endpoints/test_project_event_details.py]: I think the test `test_event_ordering` that's marked `xfail` in the original tests can probably be copied here and it should work on the snuba version.

TLDR basically if we have events `a, b, c, d, e, f, g` with the same timestamp, and we are on event `b` then our postgres approach for `prev_event` ("select 5 events with a timestamp <= the current one") would fail because it would find `c, d, e, f, g` and miss `a`, but the snuba version shouldn't have that problem

**Review by alex-hofsteede — APPROVED** [review]
This looks great, thanks for doing this implementation. 

Does this mean we will be using SnubaEvet for the event details endpoint soon? that would be awesome.

**Comment by lynnagara** [comment]
Migrated those tests to, will follow up soon with migrating event details to Snuba - seems like it's going to be quite straightforward now

File: src/sentry/db/models/manager.py [hunk 10]
```diff
@@ -342,6 +342,30 @@ def bind_nodes(self, object_list, *node_names):
             data = node_results.get(node.id) or {}
             node.bind_data(data, ref=node.get_ref(item))
 
+    def snuba_event_from_event_id(self, id_or_event_id, project_id):
```
Comment (alex-hofsteede) [inline src/sentry/db/models/manager.py]: Oh one other thing about this.

This `EventManager` is the manager for the `Event` class (set [here](https://github.com/getsentry/sentry/blob/master/src/sentry/models/event.py#L575) ) but currently `SnubaEvent` doesn't have a manager. So the interface is kinda weird for getting a SnubaEvent, currently it would be

```
Event.objects.snuba_event_from_event_id(...)
```
We could add this same manager as the manager for `SnubaEvent` too, in which case we could use.

```
SnubaEvent.objects.snuba_event_from_event_id(...)
```
But I'm not 100% sure that we should add a django model manager to a class that has nothing to do with django,  also "snuba_event" is kinda redundant in that case. So maybe there should just be a global utility function

```
sentry.event_from_event_id(id_or_event_id, project_id, snuba_event=False)
```
That returns either type of event depending on the param?

File: src/sentry/db/models/manager.py [hunk 11]
```diff
@@ -317,6 +317,32 @@ def get_queryset(self):
...
+            })
+
+            event = Event.objects.from_event_id(id_or_event_id, project_id)
+
+            if not event:
+                return None
+
+            id_or_event_id = event.event_id
+
+        return SnubaEvent.get_event(project_id, id_or_event_id)
```
Comment (alex-hofsteede) [inline src/sentry/db/models/manager.py]: We could in theory move the code for `get_event` into this method as well I guess, but it's probably fine as it is for now.

**Review by alex-hofsteede — APPROVED** [review]
Looks good
